### PR TITLE
suport MKL ml 02/05 new feature

### PR DESCRIPTION
**Pull request by glingyan**
1. support MXNet including padding pooling
2. verified googlnet-bn training

Signed-off-by: lingyan <lingyan.guo@intel.com>

### CODE REVIEW

**Comment by glingyan** [comment]
(mxnet_env) lingyan@aocl-server:~/intel_mxnet/mxnet$ python tests/python/gpu/test_operator_gpu.py 
^[[1;7H[14:26:20] src/operator/./cudnn_convolution-inl.h:55: Running performance tests to find the best convolution algorithm, this can take a while... (setting env variable MXNET_CUDNN_AUTOTUNE_DEFAULT to 0 to disable)
MKL Build:20170209
[14:26:21] src/operator/convolution.cc:40: MKLConvolutionOp Skip MKL optimization
[14:26:21] src/operator/convolution.cc:40: MKLConvolutionOp Skip MKL optimization
[14:26:21] src/operator/pooling.cu:35: Sum pooling is not supported by cudnn, MxNet sum pooling is applied.
[14:26:21] src/operator/pooling.cu:35: Sum pooling is not supported by cudnn, MxNet sum pooling is applied.
[14:26:21] src/operator/pooling.cu:35: Sum pooling is not supported by cudnn, MxNet sum pooling is applied.
[14:26:21] src/operator/pooling.cc:35: MKLPoolingOp Skip MKL optimization
[14:26:21] src/operator/pooling.cc:35: MKLPoolingOp Skip MKL optimization

**Comment by glingyan** [comment]
(mxnet_env) lingyan@aocl-server:~/intel_mxnet/mxnet$ python tests/python/unittest/test_operator.py 
MKL Build:20170209
[14:13:19] src/operator/convolution.cc:40: MKLConvolutionOp Skip MKL optimization
[14:13:19] src/operator/convolution.cc:40: MKLConvolutionOp Skip MKL optimization
[14:13:19] src/operator/convolution.cc:40: MKLConvolutionOp Skip MKL optimization
[14:13:19] src/operator/convolution.cc:40: MKLConvolutionOp Skip MKL optimization
[14:13:19] src/operator/convolution.cc:40: MKLConvolutionOp Skip MKL optimization
[14:13:19] src/operator/convolution.cc:40: MKLConvolutionOp Skip MKL optimization
[14:13:19] src/operator/convolution.cc:40: MKLConvolutionOp Skip MKL optimization
[14:13:19] src/operator/convolution.cc:40: MKLConvolutionOp Skip MKL optimization
[14:13:19] src/operator/convolution.cc:40: MKLConvolutionOp Skip MKL optimization
[14:13:19] src/operator/convolution.cc:40: MKLConvolutionOp Skip MKL optimization
[14:13:19] src/operator/convolution.cc:40: MKLConvolutionOp Skip MKL optimization
[14:13:19] src/operator/convolution.cc:40: MKLConvolutionOp Skip MKL optimization
[14:13:19] src/operator/convolution.cc:40: MKLConvolutionOp Skip MKL optimization
[14:13:19] src/operator/convolution.cc:40: MKLConvolutionOp Skip MKL optimization
[14:13:19] src/operator/convolution.cc:40: MKLConvolutionOp Skip MKL optimization
[14:13:19] src/operator/convolution.cc:40: MKLConvolutionOp Skip MKL optimization
[14:13:19] src/operator/convolution.cc:40: MKLConvolutionOp Skip MKL optimization
[14:13:19] src/operator/convolution.cc:40: MKLConvolutionOp Skip MKL optimization
[14:13:19] src/operator/convolution.cc:40: MKLConvolutionOp Skip MKL optimization
[14:13:19] src/operator/convolution.cc:40: MKLConvolutionOp Skip MKL optimization
[14:13:19] src/operator/convolution.cc:40: MKLConvolutionOp Skip MKL optimization
[14:13:19] src/operator/convolution.cc:40: MKLConvolutionOp Skip MKL optimization
[14:13:19] src/operator/convolution.cc:40: MKLConvolutionOp Skip MKL optimization
[14:13:19] src/operator/convolution.cc:40: MKLConvolutionOp Skip MKL optimization
[14:13:19] src/operator/convolution.cc:40: MKLConvolutionOp Skip MKL optimization
[14:13:19] src/operator/convolution.cc:40: MKLConvolutionOp Skip MKL optimization
[14:13:19] src/operator/convolution.cc:40: MKLConvolutionOp Skip MKL optimization
[14:13:19] src/operator/convolution.cc:40: MKLConvolutionOp Skip MKL optimization
[14:13:19] src/operator/convolution.cc:40: MKLConvolutionOp Skip MKL optimization
[14:13:19] src/operator/convolution.cc:40: MKLConvolutionOp Skip MKL optimization
[14:13:19] src/operator/convolution.cc:40: MKLConvolutionOp Skip MKL optimization
[14:13:19] src/operator/convolution.cc:40: MKLConvolutionOp Skip MKL optimization
[14:13:19] src/operator/convolution.cc:40: MKLConvolutionOp Skip MKL optimization
[14:13:19] src/operator/convolution.cc:40: MKLConvolutionOp Skip MKL optimization
[14:13:19] src/operator/convolution.cc:40: MKLConvolutionOp Skip MKL optimization
[14:13:19] src/operator/convolution.cc:40: MKLConvolutionOp Skip MKL optimization
[14:13:19] src/operator/convolution.cc:40: MKLConvolutionOp Skip MKL optimization
[14:13:19] src/operator/convolution.cc:40: MKLConvolutionOp Skip MKL optimization
[14:13:19] src/operator/convolution.cc:40: MKLConvolutionOp Skip MKL optimization
[14:13:19] src/operator/convolution.cc:40: MKLConvolutionOp Skip MKL optimization
[14:13:19] src/operator/tensor/./matrix_op-inl.h:158: Using target_shape will be deprecated.
Traceback (most recent call last):
  File "tests/python/unittest/test_operator.py", line 2775, in <module>
    test_deconvolution()
  File "tests/python/unittest/test_operator.py", line 750, in test_deconvolution
    pad                 = (1,1)
  File "tests/python/unittest/test_operator.py", line 661, in check_deconvolution_forward_backward
    assert reldiff(out + args_grad_addto_npy[0], args_grad_addto[0].asnumpy()) < 1e-6
AssertionError

currently mkl do not support kAddTo mode yet, and I found I have no way to bypass

**Comment by glingyan** [comment]
(mxnet_env) lingyan@aocl-server:~/intel_mxnet/mxnet/example/image-classification$ python train_cifar10.py --network inception-bn --batch-size 128 --lr 0.1 --lr-factor .94 --num-epoch 1 --disp-batch 1 
INFO:root:start with arguments Namespace(batch_size=128, benchmark=0, data_nthreads=4, data_train='data/cifar10_train.rec', data_val='data/cifar10_val.rec', disp_batches=1, gpus=None, image_shape='3,28,28', kv_store='device', load_epoch=None, lr=0.1, lr_factor=0.94, lr_step_epochs='200,250', max_random_aspect_ratio=0, max_random_h=36, max_random_l=50, max_random_rotate_angle=0, max_random_s=50, max_random_scale=1, max_random_shear_ratio=0, min_random_scale=1, model_prefix=None, mom=0.9, monitor=0, network='inception-bn', num_classes=10, num_epochs=1, num_examples=50000, num_layers=110, optimizer='sgd', pad_size=4, random_crop=1, random_mirror=1, rgb_mean='123.68,116.779,103.939', test_io=0, top_k=0, wd=0.0001)
[14:14:16] src/io/iter_image_recordio.cc:221: ImageRecordIOParser: data/cifar10_train.rec, use 4 threads for decoding..
[14:14:16] src/io/iter_image_recordio.cc:221: ImageRecordIOParser: data/cifar10_val.rec, use 4 threads for decoding..
MKL Build:20170209
INFO:root:Epoch[0] Batch [1]	Speed: 148.18 samples/sec	Train-accuracy=0.136719
INFO:root:Epoch[0] Batch [2]	Speed: 176.43 samples/sec	Train-accuracy=0.203125
INFO:root:Epoch[0] Batch [3]	Speed: 250.65 samples/sec	Train-accuracy=0.132812
...........................
INFO:root:Epoch[0] Time cost=302.834
INFO:root:Epoch[0] Validation-accuracy=0.600672

File: prepare_mkl.sh [hunk 1]
```diff
@@ -68,8 +68,8 @@ if [ ! -d "$HOME_MKL" ]; then
 fi
 MXNET_ROOT=`dirname $0`
 OMP=0 
-VERSION_MATCH=20161123
-ARCHIVE_BASENAME=mklml_lnx_2017.0.2.20161122.tgz
+VERSION_MATCH=20170210
+ARCHIVE_BASENAME=mklml_lnx_2017.0.2.20170209.tgz
```
Comment (piiswrong) [inline prepare_mkl.sh]: Does user need to change LD_LIBRARY_PATH?

File: src/operator/mkl/mkl_elementwise_copy-inl.h [hunk 2]
```diff
@@ -0,0 +1,69 @@
+/*******************************************************************************
```
Comment (piiswrong) [inline src/operator/mkl/mkl_elementwise_copy-inl.h]: why rename the file?

File: prepare_mkl.sh (hunk 1 above)
Comment (glingyan) [inline prepare_mkl.sh]: Makefile changes to allow use of MKL or MKL ML libs with MKL BLAS feature already complete, I will ask my colleague to commit

File: src/operator/mkl/mkl_elementwise_copy-inl.h (hunk 2 above)
Comment (glingyan) [inline src/operator/mkl/mkl_elementwise_copy-inl.h]: have new implementation of element wise , but will waiting for next mklml release to full support it

File: prepare_mkl.sh (hunk 1 above)
Comment (piiswrong) [inline prepare_mkl.sh]: I mean is this going to be extracted to mxnet/mklml_lnx_2017.0.2.20170209/ and force the users to add it into path?

Comment (glingyan) [inline prepare_mkl.sh]: yes,  it is the same with before
1.  MKLML_ROOT=/usr/local,
2. Set LD_LIBRARY_PATH=$LD_LIBRARY_PATH:$MKLML_ROOT/lib
new feature will be  review & commit by me in about 1 hous

Comment (piiswrong) [inline prepare_mkl.sh]: Every time mkldnn updates the user need to reset LD_LIBRARY_PATH. Its a lot of confusion.
 Please change it to a permanat address by making a symbolic link mxnet/mklmkl_lnx/

Comment (glingyan) [inline prepare_mkl.sh]: no .
MKLML_ROOT=/usr/local is fixed, difference mkl version will use the same MKLML_ROOT

File: src/operator/mkl/mkl_elementwise_copy-inl.h (hunk 2 above)
Comment (piiswrong) [inline src/operator/mkl/mkl_elementwise_copy-inl.h]: So currently it falls back to mxnet impl?

Comment (glingyan) [inline src/operator/mkl/mkl_elementwise_copy-inl.h]: yes, I do not touch src/operator/tensor/elemwise_sum.h

Comment (glingyan) [inline src/operator/mkl/mkl_elementwise_copy-inl.h]: mxnet use 
1. ElementWiseSumCompute for tradintion elementwise forward
2. IdentityCompute for tradintion elementwise backward
both do not touch yet

File: prepare_mkl.sh (hunk 1 above)
Comment (glingyan) [inline prepare_mkl.sh]: mailed to you some explain of LD_LIBRARY_PATH, please check your company mailbox

**Pull request merged**